enhancements/machine-config/mco-network-configuration.md

```diff
@@ -0,0 +1,290 @@
+
+#### L2 networking for Pods (Telco, CNF)
+
+As an OpenShift administrator working in Telco, I need to configure bondings and ad-hoc VLANs on day-2. Macvlan ifaces will be attached to these vlans and plugged into containers.
+
+#### Declarative container native API
+
+As an OpenShift administrator using a Baremetal platform, I want to have a declarative container native API for network interface configuration rather than have to put ifcfg/NM keyfiles.
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
+
+### Risks and Mitigations
+
+#### Not rebooting could hide errors in network configuration
+
+A NetworkManager profile is configured with certain settings (like connection.autoconnect=yes). See those settings (the content of the profiles) with `nmcli connection show "$PROFILE"`. The profile is persisted to disk (in the file you see in `nmcli -f all connection`). When NM starts, it loads the profiles from disk. NetworkManager will try to automatically connect/activate profiles that are configured to autoconnect.
+
+There is no guarantee that a reboot will succeed to connect to the network. e.g. the DHCP might simply not reply. But in general, if you have a profile (persisted to disk), and the profile is set to autoconnect, and you reboot, there is a reasonable expectation that it should work.
+
+If it doesn't come up on reboot, it's likely an environment issue (e.g. cable unplugged) or a real bug (that will be fixed by NetworkManager team).
+
+## Design Details
+
+#### Option C
+
+These configuration files would be handled via MCO, for the Baremetal platform only.
+
+In /etc/NetworkManager/conf.d/99-kni.conf we add:
+[keyfile]
+path=/etc/NetworkManager/system-connections-merged
+
+In /etc/tmpfiles.d add a file to specify:
+D /tmp/nm-system-connections 0755 root root
+D /tmp/nm-system-connections-work 0755 root root
+
+Create a .mount systemd unit to mount the overlay:
+/etc/NetworkManager/system-connections-merged
+overlay lowerdir=/etc/NetworkManager/system-connections,upperdir=/tmp/nm-system-connections,workdir=/tmp/nm-system-connections-work
+
+All the changes that nmstate will do will end up affecting just /tmp/nm-system-connections so once the machine reboots those will be gone, and only what's on /etc/NetworkManager/system-connections , which is where people would put their keyfiles with MachineConfig would remain.
+
+When kubelet would start, it would start kubernetes-nmstate again, we'd see that NNS doesn't match the NNCP that we need to apply again.
+
+The changes that knmstate + nmstate will do will happen without reboot and since they only get written to the tmpfs won't affect the next boot.
```

Comment:
In 4.6 onwards, ovn-kubernetes will use shared-gateway mode which means primary node interface will be added as a slave of ovs bridge, in such case, if vlan/bond are configured on primary node interface via this declarative API, will knmstate + nmstate be able to trigger ovs restart or re-adding updated vlan/bond interface to ovs bridge? or is it expected that the primary node interface configuration should be configured via MCO directly?

Reply:
In 4.6, the expectation is that if a node has a "complicated" network setup and is using ovn-kubernetes, then the administrator has to correctly configure the network _for ovn-kubernetes_, meaning they are responsible for setting up a correctly-configured `br-ex` containing (directly or indirectly) the node's ethernet interface(s).

Ideally, in the future, I(*) would like it to be the case that the administrator could simply describe the network setup they wanted independent of what ovn-kubernetes needs, and then some part of OCP would translate that to reflect what ovn-kubernetes needs. eg, the admin would declare "I want a VLAN with id 10 on top of a bond of eth0 and eth1", and we would automatically translate that to "I want an OVS bridge named `br-ex` containing a VLAN with id 10 on top of a bond of `eth0` and `eth1`". This is easier to do with nmstate configs than with raw NM connection files, but it's still not simply a matter of "add a new element and make the existing YAML be a child of that element" and maybe this won't actually make sense. (Notably, in the non-OVS case, "VLAN on top of a bond" involves creating kernel `bond` and `vlan` type interfaces, while in the OVS case it involves creating records within the ovsdb but no additional kernel interfaces, so you can't just create the interface as described and then move it onto the OVS bridge.)

(*) It is not yet clear if this is just me or if there will be consensus on this approach

Reply:
OK I really think we need to type up even a proto-enhancement for this ovs-by-default change.  Or failing that something we can dump into https://github.com/openshift/enhancements/blob/master/enhancements/host-network-configuration.md perhaps.

Reply:
> 
> Ideally, in the future, I(*) would like it to be the case that the administrator could simply describe the network setup they wanted independent of what ovn-kubernetes needs, and then some part of OCP would translate that to reflect what ovn-kubernetes needs. eg, the admin would declare "I want a VLAN with id 10 on top of a bond of eth0 and eth1", and we would automatically translate that to "I want an OVS bridge named `br-ex` containing a VLAN with id 10 on top of a bond of `eth0` and `eth1`". This is easier to do with nmstate configs than with raw NM connection files, but it's still not simply a matter of "add a new element and make the existing YAML be a child of that element" and maybe this won't actually make sense. (Notably, in the non-OVS case, "VLAN on top of a bond" involves creating kernel `bond` and `vlan` type interfaces, while in the OVS case it involves creating records within the ovsdb but no additional kernel interfaces, so you can't just create the interface as described and then move it onto the OVS bridge.)

In the OVS case, there could be two different bonds:
1) ovs bond
2) linux bond: used when enabling ovs hw offload with VF LAG
Ideally, we'd like to see the two bonds be supported so that user doesn't have to switch the network interfaces(MCO or this declarative network API) per feature configuration (btw, we are targeting ovs hw offload feature with linux bond in 4.7.)

> 
> (*) It is not yet clear if this is just me or if there will be consensus on this approach

It's not clear if the primary node interface configuration will be supported by this proposal as there is no explicit mention of ovs (whether it's sdn/ovn).

Reply:
To be very clear... This is for day 2 configuration. And is not meant to configure/control the primary interface/bond/vlan. There is a separate effort for that.

Reply:
> To be very clear... This is for day 2 configuration. And is not meant to configure/control the primary interface/bond/vlan. There is a separate effort for that.

@bcrochet could you list this as a non-goal in this proposal and link to the separate effort for primary interface configuration?